src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Types/RoFunctionPointerType.cs

```diff
@@ -0,0 +1,265 @@
+
+        private volatile string? _toString;
+        private volatile Type[]? _lazyCallingConventions;
+        private bool? _lazyIsUnmanaged;
+
+        private Type[] CallingConventions => _lazyCallingConventions ??= ComputeCallingConventions();
+        private string GetToString() => _toString ??= ComputeToString();
+
+        internal RoFunctionPointerType(EcmaModule module, MethodSignature<RoType> signature)
+        {
+            Debug.Assert(module != null);
+
+            _module = module;
+            _signature = signature;
+            _returnInfo = new RoFunctionPointerParameterInfo(signature.ReturnType);
+
+            ImmutableArray<RoType> parameters = signature.ParameterTypes;
+            RoFunctionPointerParameterInfo[] parameterInfos = new RoFunctionPointerParameterInfo[parameters.Length];
+            for (int i = 0; i < parameters.Length; i++)
+            {
+                parameterInfos[i] = new RoFunctionPointerParameterInfo(parameters[i]);
+            }
+            _parameterInfos = parameterInfos;
+        }
+
+        public sealed override bool IsFunctionPointer => true;
+        public sealed override bool IsUnmanagedFunctionPointer => _lazyIsUnmanaged ??= ComputeIsUnmanaged();
+        public sealed override Type[] GetFunctionPointerCallingConventions() => CallingConventions;
+        public sealed override FunctionPointerParameterInfo GetFunctionPointerReturnParameter() => _returnInfo;
+        public sealed override FunctionPointerParameterInfo[] GetFunctionPointerParameterInfos() => _parameterInfos;
+
+        private bool ComputeIsUnmanaged()
+        {
+            var _ = CallingConventions;
+            Debug.Assert(_lazyIsUnmanaged.HasValue);
+            return _lazyIsUnmanaged.Value!;
+        }
+
+        private Type[] ComputeCallingConventions()
+        {
+            SignatureCallingConvention sigCallingConvention = _signature.Header.CallingConvention;
+            Type[] customModifiers = _returnInfo.GetOptionalCustomModifiers();
+
+            bool unmanaged = false;
+            List<Type>? list = null;
+            bool foundCallingConvention = false;
+
+            for (int i = 0; i < customModifiers.Length; i++)
+            {
+                Type mod = customModifiers[i];
+                string name = mod.FullName!;
+
+                if (name.StartsWith(CallingConventionTypePrefix))
+                {
+                    list ??= new List<Type>();
+                    list.Add(mod);
+
+                    // Only consider the base calling convention, not other CallConv* Types.
+                    if (name == CallingConventionTypePrefix + sigCallingConvention.ToString())
+                    {
+                        foundCallingConvention = true;
+                        unmanaged = true;
+                    }
+                }
+            }
+
+            // Normalize the calling conventions.
+            if (!foundCallingConvention)
```

Comment:
This should be the other way around. `sigCallingConvention` should be checked first and the modopts should be read only for `SignatureCallingConvention.Unmanaged`.

That's how it is done in the runtime and Roslyn:

https://github.com/dotnet/roslyn/blob/ca8398b08b142baab1eaa5d4d89eeea61c415565/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.FunctionPointerTypeSymbolKey.cs#L18-L21
https://github.com/dotnet/runtime/blob/527f1d1e7a085395ccbf8111066f030627659d10/src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs#L1642-L1650

Reply:
In addition to normalizing the value from `GetFunctionPointerCallingConventions`, this also "normalizes" the value from `GetOptionalCustomModifiers()`, meaning for example:
- If `sigCallingConvention` is StdCall (and thus no StdCall mod opt) then we add the corresponding Stdcall modopt.
- If `sigCallingConvention` is Unmanaged and there is a modopt for StdCall, then we leave as-is.